python/rapidsmpf/rapidsmpf/shuffler.pyx

```diff
@@ -139,33 +144,47 @@ cdef class Shuffler:
         """
         # Convert python mapping to an `unordered_map`.
         cdef unordered_map[uint32_t, cpp_PackedData] _chunks
+        cdef bint _grouped = grouped  # Convert Python bool to C bool
         for pid, chunk in chunks.items():
             if not (<PackedData?>chunk).c_obj:
                 raise ValueError("PackedData was empty")
             _chunks[<uint32_t?>pid] = move(deref((<PackedData?>chunk).c_obj))
 
         with nogil:
-            deref(self._handle).insert(move(_chunks))
+            if _grouped:
+                deref(self._handle).insert_grouped(move(_chunks))
+            else:
+                deref(self._handle).insert(move(_chunks))
 
-    def insert_finished(self, uint32_t pid):
+    def insert_finished(self, pids):
         """
-        Mark a partition as finished.
+        Mark partitions as finished.
 
-        This informs the shuffler that no more chunks for the specified partition
+        This informs the shuffler that no more chunks for the specified partitions
         will be inserted.
 
         Parameters
         ----------
-        pid
-            The partition ID to mark as finished.
+        pids
+            Partition IDs to mark as finished (int or an iterable of ints).
 
         Notes
         -----
         Once a partition is marked as finished, it is considered complete and no
         further chunks will be accepted for that partition.
         """
+        cdef vector[uint32_t] _pids
+
+        if isinstance(pids, Iterable):
+            # Reserve space and populate vector
+            _pids.reserve(len(pids))
```

Comment:
Iterables don't necessarily have a length 

```python
In [3]: xs = (x for x in range(3))

In [4]: isinstance(xs, collections.abc.Iterable)
Out[4]: True

In [5]: len(xs)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 len(xs)

TypeError: object of type 'generator' has no len()
```

Maybe convert it to a Python list first, and then https://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#standard-library suggests that this should work:

```
cdef vector[uint32_t] _pids = pids_list
```

Reply:
@TomAugspurger good catch. I'll remove that `reserve`. I think that should be good enough. WDYT?

Reply:
I'd do this if it works:

```python
cdef vector[uint32_t] _pids

if isinstance(pids, Iterable):
    pids_list = list(pids)
    _pids = pids
```

Otherwise, skipping the reserve is probably fine. I wouldn't expect this to take much time regardless.

Reply:
It seems `cdef vector[int] vect = range(1, 10, 2)`  is also [allowed](https://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#standard-library), 

So, then technically we should be able to do, `_pids = pids`, isnt it?

Reply:
> So, then technically we should be able to do, _pids = pids, isnt it?

I'm not sure, but maybe! If it compiles then we're probably good to go (`range(...)` does have a length, so it's a bit different than the generator comprehension I showed above). I'm guessing / hoping that Cython does the C equivalent of `list(pids)` on whatever iterable is passed.